src/Rope.elm

```diff
@@ -0,0 +1,51 @@
+module Rope exposing (Rope(..), empty, flatFromList, fromList, one, prependTo, toList)
+
+{-| inspired by [miniBill/elm-rope](https://dark.elm.dmy.fr/packages/miniBill/elm-rope/latest/)
+-}
+
+
+type Rope a
+    = Leaf (List a)
+    | Branch (List (Rope a))
+
+
+fromList : List a -> Rope a
+fromList list =
+    Leaf list
+
+
+empty : Rope a
+empty =
+    Leaf []
```

Comment:
Considering how often we will be adding empty leafs, I wonder whether having a separate variant for empty would make sense.
In my mind, we could then potentially pattern match on that and ignore it in `prependTo`/`toListInto`

```elm
type Rope a
    = Leaf (List a)
    | Branch (List (Rope a))
    | Empty () -- So that each variant has the same arity, better for perf.

prependTo : Rope a -> Rope a -> Rope a
prependTo right left =
    case right of
      Empty () -> left
      _ ->
        case left of
            Empty () -> right
            _ -> flatFromList [ left, right ]

toListInto : List a -> Rope a -> List a
toListInto initialAcc rope =
    case rope of
        Empty -> initialAcc
         -- ...
```

I don't know if this will help, but I suspect that we will end up with a rope full with empty ropes (`Branch [ Leaf [], Leaf [], Leaf [], Branch [ Leaf [], Leaf [], Leaf [], Leaf [] ], ... ]`)

@miniBill is this something you have looked into by any chance?

Reply:
Funny you would say that. I've even created a port of elm-rope to that representation: https://github.com/lue-bird/elm-rope/tree/explicit-emptiness.
Yeah, I'll test it!

Reply:
Right, though it could reduce the amount.

Maybe just `Debug.log`ging the rope at the end of parsing `Simplify` could be interesting to see :sweat_smile: 
If it turns out to be mostly filled with comments, then it's already quite good. But my suspicion is that it's very filled with air.

Reply:
To be explicit, I'm thinking that creating this large but unnecessary data structure has a performance cost. That's a "lot" of memory to allocate and to free up during garbage collection.

Reply:
Well you were as right as it gets. But let me see if pre-filtering is faster

Reply:
yep. Is better!

Reply:
I have not in fact looked into that! My guess is that it's not a GC issue because there is a single instance of `empty`, but I could be wrong? It miiiight make sense in `fromList` to `if isEmpty x then empty else Leaf x`?

Reply:
`empty` will indeed always be a single item, with a single allocation. The problem is the `Leaf` and `Batch` that are allocated all the time, and create large lists.

I Debug.logged the resulting rope of the `Simplify.elm` module, and stored it in this Gist: https://gist.githubusercontent.com/jfmengels/c4234361ef4bf14a13b03c849867d3df/raw/44b4084980827a669fdcf3511d5fd13e6992e62c/Rope%2520emptiness.txt
(Spoiler: it's 1.1MB of text, so pretty big)

Reply:
yall are the best. I would have never believed that we can get parsing down to consistently less than 210 ms/run (with likely potential for even more speed)